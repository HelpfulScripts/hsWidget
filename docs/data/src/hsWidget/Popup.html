
                <html><style>body { overflow:hidden;}
            h1 { font-family: Arial, sans-serif; font-size: 24px; color: #44a; }
            p { margin:0; padding:0; }
            br  { margin:0; padding:0; }
            .line { 
                white-space: nowrap;
                height:16px; 
            }
            .line>span { 
                display:inline-block; background-color:#eef; height:100%; 
                margin: 0 5px 0 0; padding-right: 5px; color:#999;   
            }
            comment {color: #080;} module {color: #804;} 
            quote {color: #008;} comment>quote {color: #080;}
            .listing { margin: 10px; border: 1px solid #ccc; 
                    font-family: SFMono-Regular, Consolas, 'Liberation Mono', Menlo, Courier, monospace;
                    font-size: 14px;  
                    overflow:scroll;
                    height:90%;
            }
            code { padding: 5px 0;}</style></html><body>
                <h1>Popup.ts</h1>
                <div class='listing'><code><p id=1 class="line"><span>&nbsp;&nbsp;&nbsp;1</span><comment>/**</comment></p>
<p id=2 class="line"><span>&nbsp;&nbsp;&nbsp;2</span><comment>&nbsp;*&nbsp;#&nbsp;Popup&nbsp;Widget</comment></p>
<p id=3 class="line"><span>&nbsp;&nbsp;&nbsp;3</span><comment>&nbsp;*&nbsp;opens&nbsp;a&nbsp;popup&nbsp;on&nbsp;an&nbsp;armed&nbsp;Vnode.&nbsp;</comment></p>
<p id=4 class="line"><span>&nbsp;&nbsp;&nbsp;4</span><comment>&nbsp;*&nbsp;</comment></p>
<p id=5 class="line"><span>&nbsp;&nbsp;&nbsp;5</span><comment>&nbsp;*&nbsp;###&nbsp;Profile</comment></p>
<p id=6 class="line"><span>&nbsp;&nbsp;&nbsp;6</span><comment>&nbsp;*&nbsp;invoked&nbsp;as&nbsp;`m(Popup,&nbsp;{})`.&nbsp;`Popup`&nbsp;uses&nbsp;fixed&nbsp;positioning,&nbsp;so&nbsp;it&nbsp;can&nbsp;be&nbsp;added&nbsp;to&nbsp;the&nbsp;</comment></p>
<p id=7 class="line"><span>&nbsp;&nbsp;&nbsp;7</span><comment>&nbsp;*&nbsp;rendering&nbsp;tree&nbsp;anywhere&nbsp;except&nbsp;as&nbsp;a&nbsp;direct&nbsp;`content`&nbsp;element&nbsp;in&nbsp;`Layout`&nbsp;(where&nbsp;it&nbsp;would&nbsp;interfere&nbsp;with&nbsp;`Layout<quote>'s`&nbsp;</comment></p>
<p id=8 class="line"><span>&nbsp;&nbsp;&nbsp;8</span><comment>&nbsp;*&nbsp;rendering&nbsp;logic).&nbsp;Only&nbsp;a&nbsp;single&nbsp;`Popup`&nbsp;is&nbsp;needed&nbsp;per&nbsp;app.</comment></p>
<p id=9 class="line"><span>&nbsp;&nbsp;&nbsp;9</span><comment>&nbsp;*&nbsp;</comment></p>
<p id=10 class="line"><span>&nbsp;&nbsp;10</span><comment>&nbsp;*&nbsp;To&nbsp;arm&nbsp;a&nbsp;vnode,&nbsp;call&nbsp;`Popup.arm`&nbsp;which&nbsp;will&nbsp;return&nbsp;a&nbsp;`Mithril`&nbsp;attribute&nbsp;object:</comment></p>
<p id=11 class="line"><span>&nbsp;&nbsp;11</span><comment>&nbsp;*&nbsp;```</comment></p>
<p id=12 class="line"><span>&nbsp;&nbsp;12</span><comment>&nbsp;*&nbsp;m('</quote>.myContent<quote>',&nbsp;Popup.arm(popupContent),&nbsp;myContent);</comment></p>
<p id=13 class="line"><span>&nbsp;&nbsp;13</span><comment>&nbsp;*&nbsp;```</comment></p>
<p id=14 class="line"><span>&nbsp;&nbsp;14</span><comment>&nbsp;*&nbsp;</comment></p>
<p id=15 class="line"><span>&nbsp;&nbsp;15</span><comment>&nbsp;*&nbsp;###&nbsp;Attributes&nbsp;(node.attrs):</comment></p>
<p id=16 class="line"><span>&nbsp;&nbsp;16</span><comment>&nbsp;*&nbsp;none</comment></p>
<p id=17 class="line"><span>&nbsp;&nbsp;17</span><comment>&nbsp;*&nbsp;</comment></p>
<p id=18 class="line"><span>&nbsp;&nbsp;18</span><comment>&nbsp;*&nbsp;###&nbsp;Example</comment></p>
<p id=19 class="line"><span>&nbsp;&nbsp;19</span><comment>&nbsp;*&nbsp;<example></comment></p>
<p id=20 class="line"><span>&nbsp;&nbsp;20</span><comment>&nbsp;*&nbsp;<file&nbsp;name='</quote>script.js<quote>'></comment></p>
<p id=21 class="line"><span>&nbsp;&nbsp;21</span><comment>&nbsp;*&nbsp;let&nbsp;dismissals&nbsp;=&nbsp;0;</comment></p>
<p id=22 class="line"><span>&nbsp;&nbsp;22</span><comment>&nbsp;*&nbsp;let&nbsp;trigger;</comment></p>
<p id=23 class="line"><span>&nbsp;&nbsp;23</span><comment>&nbsp;*&nbsp;m.mount(root,&nbsp;{view:&nbsp;()&nbsp;=>&nbsp;m('</quote>.hs-white<quote>',&nbsp;[</comment></p>
<p id=24 class="line"><span>&nbsp;&nbsp;24</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;m('</quote>h4<quote>',&nbsp;hsWidget.Popup.arm(`popup&nbsp;content`),&nbsp;`hover&nbsp;over&nbsp;me`),</comment></p>
<p id=25 class="line"><span>&nbsp;&nbsp;25</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;m(hsWidget.Popup,&nbsp;{})</comment></p>
<p id=26 class="line"><span>&nbsp;&nbsp;26</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;])</comment></p>
<p id=27 class="line"><span>&nbsp;&nbsp;27</span><comment>&nbsp;*&nbsp;});</comment></p>
<p id=28 class="line"><span>&nbsp;&nbsp;28</span><comment>&nbsp;*&nbsp;</file></comment></p>
<p id=29 class="line"><span>&nbsp;&nbsp;29</span><comment>&nbsp;*&nbsp;</example></comment></p>
<p id=30 class="line"><span>&nbsp;&nbsp;30</span><comment>&nbsp;*/</comment></p>
<p id=31 class="line"><span>&nbsp;&nbsp;31</span></p>
<p id=32 class="line"><span>&nbsp;&nbsp;32</span>&nbsp;<comment>/**&nbsp;*/</comment></p>
<p id=33 class="line"><span>&nbsp;&nbsp;33</span>import&nbsp;{&nbsp;Log&nbsp;}&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;from&nbsp;'</quote>hsutil<quote>';&nbsp;const&nbsp;log&nbsp;=&nbsp;new&nbsp;Log('</quote>Popup<quote>');</p>
<p id=34 class="line"><span>&nbsp;&nbsp;34</span>import&nbsp;{&nbsp;m,&nbsp;Vnode}&nbsp;&nbsp;from&nbsp;'</quote>hslayout<quote>';&nbsp;</p>
<p id=35 class="line"><span>&nbsp;&nbsp;35</span><comment>//&nbsp;import&nbsp;{&nbsp;stringify&nbsp;}&nbsp;from&nbsp;'</quote>querystring<quote>';</comment></p>
<p id=36 class="line"><span>&nbsp;&nbsp;36</span><comment></comment></p>
<p id=37 class="line"><span>&nbsp;&nbsp;37</span><comment>/**</comment></p>
<p id=38 class="line"><span>&nbsp;&nbsp;38</span><comment>&nbsp;*&nbsp;a&nbsp;`Mithril`&nbsp;node&nbsp;that&nbsp;shows&nbsp;a&nbsp;popup&nbsp;when&nbsp;triggered.</comment></p>
<p id=39 class="line"><span>&nbsp;&nbsp;39</span><comment>&nbsp;*/</comment></p>
<p id=40 class="line"><span>&nbsp;&nbsp;40</span>export&nbsp;class&nbsp;Popup&nbsp;{</p>
<p id=41 class="line"><span>&nbsp;&nbsp;41</span>&nbsp;&nbsp;&nbsp;&nbsp;<comment>/**</comment></p>
<p id=42 class="line"><span>&nbsp;&nbsp;42</span><comment>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;*&nbsp;arms&nbsp;a&nbsp;`Mithril`&nbsp;node&nbsp;by&nbsp;adding&nbsp;mouse&nbsp;listeners&nbsp;to&nbsp;the&nbsp;`attrs`&nbsp;object.&nbsp;</comment></p>
<p id=43 class="line"><span>&nbsp;&nbsp;43</span><comment>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;*&nbsp;@param&nbsp;content&nbsp;a&nbsp;`VNode`&nbsp;that&nbsp;specifies&nbsp;the&nbsp;content&nbsp;to&nbsp;show&nbsp;in&nbsp;the&nbsp;popup.</comment></p>
<p id=44 class="line"><span>&nbsp;&nbsp;44</span><comment>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;*&nbsp;@param&nbsp;attrs&nbsp;optional;&nbsp;an&nbsp;`attrs`&nbsp;object&nbsp;that&nbsp;is&nbsp;extended&nbsp;by&nbsp;mouse&nbsp;listeners.</comment></p>
<p id=45 class="line"><span>&nbsp;&nbsp;45</span><comment>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;*&nbsp;@return&nbsp;the&nbsp;extended&nbsp;`attrs`&nbsp;object</comment></p>
<p id=46 class="line"><span>&nbsp;&nbsp;46</span><comment>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;*/</comment></p>
<p id=47 class="line"><span>&nbsp;&nbsp;47</span>&nbsp;&nbsp;&nbsp;&nbsp;public&nbsp;static&nbsp;arm(content:Vnode,&nbsp;attrs:any={}):any&nbsp;{</p>
<p id=48 class="line"><span>&nbsp;&nbsp;48</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;attrs.onmouseenter&nbsp;=&nbsp;(e:any)&nbsp;=>&nbsp;Popup.instance.show(e,&nbsp;content);</p>
<p id=49 class="line"><span>&nbsp;&nbsp;49</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;attrs.onmousemove&nbsp;=&nbsp;(e:any)&nbsp;=>&nbsp;Popup.instance.move(e);</p>
<p id=50 class="line"><span>&nbsp;&nbsp;50</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;attrs.onmouseleave&nbsp;=&nbsp;()&nbsp;=>&nbsp;Popup.instance.hide();</p>
<p id=51 class="line"><span>&nbsp;&nbsp;51</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;attrs.class&nbsp;=&nbsp;`popup&nbsp;${attrs.class||'</quote><quote>'}`;</p>
<p id=52 class="line"><span>&nbsp;&nbsp;52</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;return&nbsp;attrs;</p>
<p id=53 class="line"><span>&nbsp;&nbsp;53</span>&nbsp;&nbsp;&nbsp;&nbsp;}</p>
<p id=54 class="line"><span>&nbsp;&nbsp;54</span>&nbsp;&nbsp;&nbsp;&nbsp;<comment>/**&nbsp;the&nbsp;singleton&nbsp;instance&nbsp;to&nbsp;use&nbsp;upon&nbsp;mouse&nbsp;events&nbsp;*/</comment></p>
<p id=55 class="line"><span>&nbsp;&nbsp;55</span>&nbsp;&nbsp;&nbsp;&nbsp;private&nbsp;static&nbsp;instance:Popup;</p>
<p id=56 class="line"><span>&nbsp;&nbsp;56</span></p>
<p id=57 class="line"><span>&nbsp;&nbsp;57</span>&nbsp;&nbsp;&nbsp;&nbsp;<comment>/**&nbsp;the&nbsp;fixed&nbsp;x&nbsp;position&nbsp;of&nbsp;the&nbsp;popup&nbsp;*/</comment></p>
<p id=58 class="line"><span>&nbsp;&nbsp;58</span>&nbsp;&nbsp;&nbsp;&nbsp;x&nbsp;=&nbsp;0;</p>
<p id=59 class="line"><span>&nbsp;&nbsp;59</span>&nbsp;&nbsp;&nbsp;&nbsp;<comment>/**&nbsp;the&nbsp;fixed&nbsp;y&nbsp;position&nbsp;of&nbsp;the&nbsp;popup&nbsp;*/</comment></p>
<p id=60 class="line"><span>&nbsp;&nbsp;60</span>&nbsp;&nbsp;&nbsp;&nbsp;y&nbsp;=&nbsp;0;</p>
<p id=61 class="line"><span>&nbsp;&nbsp;61</span>&nbsp;&nbsp;&nbsp;&nbsp;<comment>/**&nbsp;state&nbsp;variable,&nbsp;true&nbsp;if&nbsp;popup&nbsp;is&nbsp;visible.&nbsp;*/</comment></p>
<p id=62 class="line"><span>&nbsp;&nbsp;62</span>&nbsp;&nbsp;&nbsp;&nbsp;showPopup&nbsp;=&nbsp;false;</p>
<p id=63 class="line"><span>&nbsp;&nbsp;63</span>&nbsp;&nbsp;&nbsp;&nbsp;<comment>/**&nbsp;state&nbsp;variable&nbsp;containing&nbsp;the&nbsp;content&nbsp;to&nbsp;show.&nbsp;*/</comment></p>
<p id=64 class="line"><span>&nbsp;&nbsp;64</span>&nbsp;&nbsp;&nbsp;&nbsp;content:Vnode&nbsp;=&nbsp;'</quote>...Popup...<quote>';</p>
<p id=65 class="line"><span>&nbsp;&nbsp;65</span></p>
<p id=66 class="line"><span>&nbsp;&nbsp;66</span>&nbsp;&nbsp;&nbsp;&nbsp;private&nbsp;show(e:any,&nbsp;content:Vnode)&nbsp;{</p>
<p id=67 class="line"><span>&nbsp;&nbsp;67</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;this.showPopup&nbsp;=&nbsp;true;</p>
<p id=68 class="line"><span>&nbsp;&nbsp;68</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;this.content&nbsp;=&nbsp;content;</p>
<p id=69 class="line"><span>&nbsp;&nbsp;69</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;this.move(e);</p>
<p id=70 class="line"><span>&nbsp;&nbsp;70</span>&nbsp;&nbsp;&nbsp;&nbsp;}</p>
<p id=71 class="line"><span>&nbsp;&nbsp;71</span>&nbsp;&nbsp;&nbsp;&nbsp;private&nbsp;move(e:any)&nbsp;{</p>
<p id=72 class="line"><span>&nbsp;&nbsp;72</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;this.x&nbsp;=&nbsp;e.x;</p>
<p id=73 class="line"><span>&nbsp;&nbsp;73</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;this.y&nbsp;=&nbsp;e.y;</p>
<p id=74 class="line"><span>&nbsp;&nbsp;74</span>&nbsp;&nbsp;&nbsp;&nbsp;}</p>
<p id=75 class="line"><span>&nbsp;&nbsp;75</span>&nbsp;&nbsp;&nbsp;&nbsp;private&nbsp;hide()&nbsp;{</p>
<p id=76 class="line"><span>&nbsp;&nbsp;76</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;this.showPopup&nbsp;=&nbsp;false;</p>
<p id=77 class="line"><span>&nbsp;&nbsp;77</span>&nbsp;&nbsp;&nbsp;&nbsp;}</p>
<p id=78 class="line"><span>&nbsp;&nbsp;78</span>&nbsp;&nbsp;&nbsp;&nbsp;view(node:Vnode)&nbsp;{</p>
<p id=79 class="line"><span>&nbsp;&nbsp;79</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Popup.instance&nbsp;=&nbsp;this;</p>
<p id=80 class="line"><span>&nbsp;&nbsp;80</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;const&nbsp;pos&nbsp;=&nbsp;xy(window.innerWidth,&nbsp;window.innerHeight,&nbsp;this.x,&nbsp;this.y);</p>
<p id=81 class="line"><span>&nbsp;&nbsp;81</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;const&nbsp;attrs&nbsp;=&nbsp;{&nbsp;style:`left:${pos[0]}px;&nbsp;top:${pos[1]}px;&nbsp;transform:&nbsp;translate(${pos[2]}%,&nbsp;${pos[3]}%);`};</p>
<p id=82 class="line"><span>&nbsp;&nbsp;82</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;return&nbsp;!this.showPopup?&nbsp;m('</quote>span')&nbsp;:&nbsp;m(`.hs_popup`,&nbsp;attrs,&nbsp;this.content);</p>
<p id=83 class="line"><span>&nbsp;&nbsp;83</span>&nbsp;&nbsp;&nbsp;&nbsp;}</p>
<p id=84 class="line"><span>&nbsp;&nbsp;84</span>}</p>
<p id=85 class="line"><span>&nbsp;&nbsp;85</span></p>
<p id=86 class="line"><span>&nbsp;&nbsp;86</span>function&nbsp;xy(w:number,&nbsp;h:number,&nbsp;x:number,&nbsp;y:number)&nbsp;{</p>
<p id=87 class="line"><span>&nbsp;&nbsp;87</span>&nbsp;&nbsp;&nbsp;&nbsp;let&nbsp;left&nbsp;=&nbsp;x;</p>
<p id=88 class="line"><span>&nbsp;&nbsp;88</span>&nbsp;&nbsp;&nbsp;&nbsp;let&nbsp;top&nbsp;=&nbsp;y;</p>
<p id=89 class="line"><span>&nbsp;&nbsp;89</span>&nbsp;&nbsp;&nbsp;&nbsp;let&nbsp;transX&nbsp;=&nbsp;50;</p>
<p id=90 class="line"><span>&nbsp;&nbsp;90</span>&nbsp;&nbsp;&nbsp;&nbsp;let&nbsp;transY&nbsp;=&nbsp;-50;</p>
<p id=91 class="line"><span>&nbsp;&nbsp;91</span>&nbsp;&nbsp;&nbsp;&nbsp;if&nbsp;(2*x&nbsp;>&nbsp;w)&nbsp;{&nbsp;</p>
<p id=92 class="line"><span>&nbsp;&nbsp;92</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;left+=&nbsp;15;&nbsp;transX&nbsp;=&nbsp;-100;&nbsp;</p>
<p id=93 class="line"><span>&nbsp;&nbsp;93</span>&nbsp;&nbsp;&nbsp;&nbsp;}&nbsp;else&nbsp;{</p>
<p id=94 class="line"><span>&nbsp;&nbsp;94</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;left-=&nbsp;15;&nbsp;transX&nbsp;=&nbsp;0;</p>
<p id=95 class="line"><span>&nbsp;&nbsp;95</span>&nbsp;&nbsp;&nbsp;&nbsp;}</p>
<p id=96 class="line"><span>&nbsp;&nbsp;96</span>&nbsp;&nbsp;&nbsp;&nbsp;if&nbsp;(2*y&nbsp;>&nbsp;h)&nbsp;{&nbsp;</p>
<p id=97 class="line"><span>&nbsp;&nbsp;97</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;top-=&nbsp;15;&nbsp;transY&nbsp;=&nbsp;-100;&nbsp;</p>
<p id=98 class="line"><span>&nbsp;&nbsp;98</span>&nbsp;&nbsp;&nbsp;&nbsp;}&nbsp;else&nbsp;{</p>
<p id=99 class="line"><span>&nbsp;&nbsp;99</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;top+=&nbsp;15;&nbsp;transY&nbsp;=&nbsp;0;</p>
<p id=100 class="line"><span>&nbsp;100</span>&nbsp;&nbsp;&nbsp;&nbsp;}</p>
<p id=101 class="line"><span>&nbsp;101</span>&nbsp;&nbsp;&nbsp;&nbsp;return&nbsp;[left,&nbsp;top,&nbsp;transX,&nbsp;transY];</p>
<p id=102 class="line"><span>&nbsp;102</span>}</p></code></div>
                </body>
            