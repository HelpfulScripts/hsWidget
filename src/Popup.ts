/**
 * # Popup Widget
 * opens a popup on an armed Vnode. 
 * 
 * ### Profile
 * invoked as `m(Popup, {})`. `Popup` uses fixed positioning, so it can be added to the 
 * rendering tree anywhere except as a direct `content` element in `Layout` (where it would interfere with `Layout's` 
 * rendering logic). Only a single `Popup` is needed per app.
 * 
 * To arm a vnode, call `Popup.arm` which will return a `Mithril` attribute object:
 * ```
 * m('.myContent', Popup.arm(popupContent), myContent);
 * ```
 * Additional attributes can be passed into the `arm` method and will be passed on to the hosting component. For example,
 * to add a `Popup` to the component 
 * ```
 * m('.myContent', { onclick: click }, myContent);
 * ```
 * do the following:
 * ```
 * m('.myContent', Popup.arm(popupContent, {onclick: click}), myContent);
 * ```
 * 
 * ### Attributes (node.attrs):
 * none
 * 
 * ### Example
 * <example>
 * <file name='script.js'>
 * let dismissals = 0;
 * let trigger;
 * m.mount(root, {view: () => m('.hs-white', [
 *      m('h4', hsWidget.Popup.arm(`popup content`), `hover over me`),
 *      m(hsWidget.Popup, {})
 *    ])
 * });
 * </file>
 * </example>
 */

 /** */
import { Log }      from 'hsutil'; const log = new Log('Popup');
import m from "mithril";
type Vnode = m.Vnode<any, any>;

/**
 * a `Mithril` node that shows a popup when triggered.
 */
export class Popup {
    /**
     * arms a `Mithril` node by adding mouse listeners to the `attrs` object. 
     * @param content a `VNode` that specifies the content to show in the popup.
     * @param attrs optional; an `attrs` object that is extended by mouse listeners.
     * @return the extended `attrs` object
     */
    public static arm(content:m.Children, attrs:any={}):any {
        if (content) {
            attrs.onmouseenter = (e:any) => Popup.instance.show(e, content);
            attrs.onmousemove = (e:any) => Popup.instance.move(e);
            attrs.onmouseleave = () => Popup.instance.hide();
            attrs.class = `popup ${attrs.class||''}`;
        }
        return attrs;
    }
    /** the singleton instance to use upon mouse events */
    private static instance:Popup;

    /** the fixed x position of the popup */
    x = 0;
    /** the fixed y position of the popup */
    y = 0;
    /** state variable, true if popup is visible. */
    showPopup = false;
    /** state variable containing the content to show. */
    content:m.Children = '...Popup...';

    private show(e:any, content:m.Children) {
        this.showPopup = true;
        this.content = content;
        this.move(e);
    }
    private move(e:any) {
        this.x = e.x;
        this.y = e.y;
    }
    private hide() {
        this.showPopup = false;
    }
    view(node:Vnode) {
        Popup.instance = this;
        const pos = xy(window.innerWidth, window.innerHeight, this.x, this.y);
        const attrs = { style:`left:${pos[0]}px; top:${pos[1]}px; transform: translate(${pos[2]}%, ${pos[3]}%);`};
        return !this.showPopup? m('span') : m(`.hs_popup`, attrs, this.content);
    }
}

function xy(w:number, h:number, x:number, y:number) {
    let left = x;
    let top = y;
    let transX = 50;
    let transY = -50;
    if (2*x > w) { 
        left+= 15; transX = -100; 
    } else {
        left-= 15; transX = 0;
    }
    if (2*y > h) { 
        top-= 15; transY = -100; 
    } else {
        top+= 15; transY = 0;
    }
    return [left, top, transX, transY];
}