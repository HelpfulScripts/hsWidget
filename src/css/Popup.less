.hs_popup {
    position: fixed;
    padding: 3px 5px;
    z-index: 999;
    background-color:#eee;
    border-radius: 5px; 
    border: 1px solid #aaf;
    width: auto;
    height: auto;
    font-size: 14px;
}

