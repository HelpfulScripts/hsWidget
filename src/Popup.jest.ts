import m from "mithril";
import { Popup }    from './Popup';
const root = window.document.createElement("div");


describe('EditCheckbox', () => {
    beforeAll(()=>{
        m.mount(root, {view: () => m('.hs-white', [
            m('h4', Popup.arm(`popup content`), `hover over me`),
            m(Popup, {})
          ])
       });
    });

    it('should match snapshot', () => {
        expect(root).toMatchSnapshot();
    });
});