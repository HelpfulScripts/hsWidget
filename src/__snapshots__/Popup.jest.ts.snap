// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditCheckbox should match snapshot 1`] = `
<div>
  <div
    class="hs-white"
  >
    <h4
      class="popup "
    >
      hover over me
    </h4>
    <span />
  </div>
</div>
`;
